newton.go

```diff
@@ -160,415 +73,93 @@ func (n *Newton) sign(req *http.Request) error {
-	a[1].Value = strconv.FormatInt(endDate, 10)
-
-	a[2].Key = "limit"
-	a[2].Value = strconv.Itoa(limit)
-
-	a[3].Key = "offset"
-	a[3].Value = strconv.Itoa(offset)
-
-	a[4].Key = "start_date"
-	a[4].Value = strconv.FormatInt(startDate, 10)
-
-	res, err := n.doPrivateQuery("/actions", http.MethodGet, a, "")
+	res, err := http.DefaultClient.Do(req)
 	if err != nil {
 		return nil, err
 	}
-	defer func() {
-		err := res.Body.Close()
-		if err != nil {
-			log.Printf("error:%s", err.Error())
-		}
-	}()
-	body, _ := ioutil.ReadAll(res.Body)
-	if res.StatusCode != http.StatusOK {
-		return nil, errors.New(fmt.Sprintf("request failed :: %d", res.StatusCode))
-	}
 
-	var r ActionsResp
-	err = json.Unmarshal(body, &r.Actions)
+	parsedResponse, err := n.parseResponse(res, query.GetResponse())
 	if err != nil {
 		return nil, err
 	}
 
-	return &r, nil
+	return parsedResponse, err
 }
 
-func (n *Newton) OrdersHistory(limit int, offset int, startDate int64, endDate int64, symbol string, timeInForce string) (*OrdersHistoryResp, error) {
-
-	a := make([]Args, 10)
-
-	a[1].Key = "end_date"
-	a[1].Value = strconv.FormatInt(endDate, 10)
-
-	a[2].Key = "limit"
-	a[2].Value = strconv.Itoa(limit)
-
-	a[3].Key = "offset"
-	a[3].Value = strconv.Itoa(offset)
-
-	a[4].Key = "start_date"
-	a[4].Value = strconv.FormatInt(startDate, 10)
-
-	a[5].Key = "symbol"
-	a[5].Value = symbol
-
-	a[6].Key = "time_in_force"
-	a[6].Value = timeInForce
-
-	res, err := n.doPrivateQuery("/order/history", http.MethodGet, a, "")
-	if err != nil {
-		return nil, err
-	}
+func (n *Newton) parseResponse(res *http.Response, toParseTo interface{}) (*Response, error) {
 	defer func() {
 		err := res.Body.Close()
 		if err != nil {
 			log.Printf("error:%s", err.Error())
 		}
 	}()
-	body, _ := ioutil.ReadAll(res.Body)
-	if res.StatusCode != http.StatusOK {
-		return nil, errors.New(fmt.Sprintf("request failed :: %d", res.StatusCode))
-	}
 
-	var r OrdersHistoryResp
-	err = json.Unmarshal(body, &r.OrdersHistory)
-	if err != nil {
-		return nil, err
+	parsedResponse := &Response{
+		StatusCode: res.StatusCode,
+		Body: nil,
 	}
 
-	return &r, nil
-}
-
-func (n *Newton) OpenOrders(limit int, offset int, symbol string, timeInForce string) (*OpenOrdersResp, error) {
-
-	a := make([]Args, 10)
-
-	a[1].Key = "limit"
-	a[1].Value = strconv.Itoa(limit)
-
-	a[2].Key = "offset"
-	a[2].Value = strconv.Itoa(offset)
-
-	a[3].Key = "symbol"
-	a[3].Value = symbol
-
-	a[4].Key = "time_in_force"
-	a[4].Value = timeInForce
-
-	res, err := n.doPrivateQuery("/order/history", http.MethodGet, a, "")
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		err := res.Body.Close()
-		if err != nil {
-			log.Printf("error:%s", err.Error())
-		}
-	}()
-	if res.StatusCode != http.StatusOK {
-		return nil, errors.New(fmt.Sprintf("request failed :: %d", res.StatusCode))
+	if toParseTo == nil {
+		return parsedResponse, nil
 	}
 
-	body, _ := ioutil.ReadAll(res.Body)
+	body, err := ioutil.ReadAll(res.Body)
```

Comment:
I think we should read the body first and then decide if we return a nil value and error if a body is returned and `toParseTo` is `nil`, the reason is that if the body of a request is not read properly by the client it can reset the connexion and the body is never properly freed.

Reply:
Could you elaborate on this please? I'd assume it is safe since we close response's Body in defer. Why do we need to read the body in addition to closing to ensure that it is properly freed?

Reply:
there is a special behaviour in go http client that if you close the body without reading the buffer it will reset the underlying connexion and that could be a source of performance issues. [more details in here](https://stackoverflow.com/questions/18598780/is-resp-body-close-necessary-if-we-dont-read-anything-from-the-body)

Reply:
thanks for the explanation